### PR TITLE
New time gen

### PR DESCRIPTION
**Pull request by kingjr**
Finalize and address Alex's comments, trial 2.

### CODE REVIEW

File: examples/decoding/plot_decoding_time_generalization.py [hunk 1]
```diff
@@ -82,7 +82,7 @@
 viz_vs_auditory_l = (events[np.in1d(events, (1, 3))] == 3).astype(int)
 # To make scikit-learn happy, we converted the bool array to integers
 # in the same line. This results in an array of zeros and ones:
-print np.unique(viz_vs_auditory_l)
+print("The unique classes' labels are " + str(np.unique(viz_vs_auditory_l)))
```
Comment (dengemann) [inline examples/decoding/plot_decoding_time_generalization.py]: @kingjr 

``` Python
"The unique classes' labels are %s" % np.unique(viz_vs_auditory_l)
```

**Comment by kingjr** [comment]
Several points for me to keep in mind
1. travis s not happy.
2. I still need to have a look at the decision function (or next PR?)
3. I need to avoid the too-smart approach of removing one prediction when redundant (proba, 1-proba)

**Comment by kingjr** [comment]
@dengemann I don't understand travis issue. Can you check it out?

**Comment by dengemann** [comment]
I suspect you have condition in the gat.**repr** that's bugging ... The crash occurred first when you added the repr

**Comment by kingjr** [comment]
There's still a problem when predict_type="decision_function" and classes != [0, 1], but I have the feeling this should be solved at the sklearn level.

**Comment by kingjr** [comment]
Apart from this, and travis issue, I think we're done over here. You can review it. note that I added some XXX where I wasn't sure.

**Comment by dengemann** [comment]
RuntimeError: predict_type must be "predict" or "predict_proba"

2014-12-11 14:18 GMT+01:00 J-R King notifications@github.com:

**Comment by kingjr** [comment]
I was testing the wrong testing function before pushing the script. Should be ok now. We still have an error when data.events is not binary and we're trying to use the roc scoring function, but I think this should be dealt at the sklearn level.

**Comment by kingjr** [comment]
The remaining error is on python 3.4, the default sklearn scorer seem not to have the attribute "func_name". what do we do?

**Comment by agramfort** [comment]
what version of sklearn does this code require? you can protect the test to only run for latest version.

**Comment by kingjr** [comment]
On my computer I'm using 0.14.1. Do you have an example on how to protect for specific versions?

**Comment by dengemann** [comment]
JR, study mne/utils.py a bit, you will find many many examples of version
handling. E.g. related to sklearn. IIRC we also require certain versions
for ICA. It would be good though to figure out how one can use the same
functionality in python 3.4 using sklearn.

2014-12-11 22:09 GMT+01:00 J-R King notifications@github.com:

File: mne/decoding/time_gen.py [hunk 2]
```diff
@@ -14,6 +14,56 @@
...
+        'step' : float
+            Duration separating the start of subsequent classifiers (in
+            seconds). By default, equals one time sample.
+        'length' : float
+            Duration of each classifier (in seconds). By default, equals one
+            time sample.
+        If None, empty dict. Defaults to None."""
+    def __init__(self):
+        # XXX JRK: Is there something to be done here?
+        pass
```
Comment (dengemann) [inline mne/decoding/time_gen.py]: not really but you have to add the data somewhere. You can probablyjust not re define the init

File: mne/decoding/time_gen.py [hunk 3]
```diff
@@ -98,7 +157,7 @@ def __init__(self, cv=5, clf=None, train_times=None, predict_type='predict',
         # Store parameters in object
         self.cv = cv
         # Define training sliding window
-        self.train_times = {} if train_times is None else train_times
+        self.train_times = DecodingTime() if train_times is None else train_times
```
Comment (dengemann) [inline mne/decoding/time_gen.py]: It seems you need to populate DecodingTime with train_times etc to make it work.

**Comment by dengemann** [comment]
> The remaining error is on python 3.4, the default sklearn scorer seem not to have the attribute "func_name". what do we do?

just never rely on such attributes but use the repr '%s' % function and test for substrings. alternatively use the inspect module

**Comment by dengemann** [comment]
@kingjr looking at travis the error seems to be somewhat different. we get the ValueError because predict_type seems to get an invalid value somewhere

**Comment by kingjr** [comment]
Are you sure? I fixed the func_name / python 3 issue, and Trevis seems fine. 

I think I m done over here, reviews welcome.

File: mne/decoding/tests/test_time_gen.py [hunk 4]
```diff
@@ -66,13 +66,22 @@ def test_generalization_across_time():
 
     # Test default running
     gat = GeneralizationAcrossTime()
+    print(gat)  # check __repr__
     gat.fit(epochs)
+    print(gat)  # check __repr__
     gat.predict(epochs)
+    print(gat)  # check __repr__
     gat.score(epochs)
+    print(gat)  # check __repr__
```
Comment (dengemann) [inline mne/decoding/tests/test_time_gen.py]: don't print here

get the string instead and test it.

`output = '%s' % gat`

File: mne/decoding/tests/test_time_gen.py [hunk 5]
```diff
@@ -66,13 +66,22 @@ def test_generalization_across_time():
...
     gat.fit(epochs, y=epochs.events[:, 2])
     gat.score(epochs, y=epochs.events[:, 2])
+    assert_true("accuracy_score" in gat.scorer_.__repr__())    
     epochs2 = epochs.copy()
 
+    # check DecodingTime class
+    print(gat.train_times)
+    print(gat.test_times_)
```
Comment (dengemann) [inline mne/decoding/tests/test_time_gen.py]: same here

File: mne/decoding/tests/test_time_gen.py [hunk 6]
```diff
@@ -138,16 +147,37 @@ def test_generalization_across_time():
...
+
+    # test various predict_type
+    gat = GeneralizationAcrossTime(clf=svc, predict_type="predict_proba")
+    gat.fit(epochs)
+    gat.predict(epochs)
+    # check that 2 class probabilistic estimates are [p, 1-p]
+    assert_true(gat.y_pred_.shape[3] == 2)  
+    gat.score(epochs)
+    # check that continuous prediction leads to AUC rather than accuracy
+    assert_true("roc_auc_score" in gat.scorer_.__repr__())
```
Comment (dengemann) [inline mne/decoding/tests/test_time_gen.py]: don't call `__repr__` directly:

`assert_true("roc_auc_score" in '%s' % gat.score)`

File: mne/decoding/tests/test_time_gen.py [hunk 7]
```diff
@@ -138,16 +147,37 @@ def test_generalization_across_time():
...
+    # XXX Sklearn doesn't like non-binary inputs. We could binarize the data, 
+    # or change Sklearn default behavior
+    epochs.events[:, 2][epochs.events[:, 2] == 3] = 0
+    gat.fit(epochs)
+    gat.predict(epochs)
+    # check that 2 class non-probabilistic continuous estimates are [distance]
+    assert_true(gat.y_pred_.shape[3] == 1)
+    gat.score(epochs)
+    # check that continuous prediction leads to AUC rather than accuracy
+    assert_true("roc_auc_score" in gat.scorer_.__repr__())
```
Comment (dengemann) [inline mne/decoding/tests/test_time_gen.py]: same here. Also empty line is missing at the end of the file

File: mne/decoding/time_gen.py [hunk 8]
```diff
@@ -14,6 +14,53 @@
 from ..utils import logger, verbose, deprecated
 from ..io.pick import channel_type, pick_types
 
+class DecodingTime(dict):
+    # XXX JRK: is this class defined in the right place?
+    # XXX JRK: Are the following comments adequate?
```
Comment (dengemann) [inline mne/decoding/time_gen.py:19]: XXX still relevant?

File: mne/decoding/time_gen.py [hunk 9]
```diff
@@ -98,7 +155,8 @@ def __init__(self, cv=5, clf=None, train_times=None, predict_type='predict',
         # Store parameters in object
         self.cv = cv
         # Define training sliding window
-        self.train_times = {} if train_times is None else train_times
+        self.train_times = DecodingTime() if train_times is None \
+                                          else DecodingTime(train_times)
```
Comment (dengemann) [inline mne/decoding/time_gen.py:159]: is this correct? should DecodingTime be populated like a dict?

Comment (kingjr) [inline mne/decoding/time_gen.py:159]: This returns a DecodingTime dic, isn't that what we want? Else, can you make an example?

File: mne/decoding/time_gen.py (hunk 8 above)
Comment (kingjr) [inline mne/decoding/time_gen.py:19]: Yes, I'm not clear whether we can make the **repr** output less shaky. Is there a specific format we can follow across objects?

File: mne/decoding/time_gen.py (hunk 9 above)
Comment (dengemann) [inline mne/decoding/time_gen.py:159]: I did not fully unpack the end of the line :) this code works because train times is a dict already, got it

**Comment by kingjr** [comment]
@dengemann I don't have have python 3, and thus don't manage to replicate the error. Do you know how I could track it?

**Comment by dengemann** [comment]
create an env for python 3, it's easy: http://continuum.io/blog/anaconda-python-3

**Comment by dengemann** [comment]
@kingjr before digging deeper cany you push -f to see how travis rebuilds?

**Comment by kingjr** [comment]
I try, but I don't manage. I rebased, but it say that my branch is still behind the remote counterpart:
$ git push kingjr new_time_gen 
To git@github.com:kingjr/mne-python.git
 ! [rejected]        new_time_gen -> new_time_gen (non-fast-forward)
error: failed to push some refs to 'git@github.com:kingjr/mne-python.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

$ git co master
error: Your local changes to the following files would be overwritten by checkout:
    examples/decoding/plot_decoding_time_generalization.py
    mne/decoding/tests/test_time_gen.py
Please, commit your changes or stash them before you can switch branches.
Aborting

$ git pull upstream master
From https://github.com/mne-tools/mne-python
- branch            master     -> FETCH_HEAD
  Already up-to-date.

**Comment by dengemann** [comment]
did you complete the rebase? did you try pushing using -f

**Pull request merged**